File: Haskell/ue_7/ue_7.java
/* Codevorlage zum 7. Foliensatz
   Alternative Programmierparadigmen 2025
   Prof. Baltasar Trancón Widemann */

import java.lang.Math;

abstract class Expr {

  public abstract int eval();

  @Override
  public abstract String toString();

  protected String toStringNeg() {
    return "(" + toString() + ")";
  }

  public abstract boolean hasZero();

  public abstract int noNegConstants();

}

class Constant extends Expr {

  private final int value;

  public Constant(int value) {
    this.value = value;
  }

  @Override
  public int eval() {
    return value;
  }

  @Override
  public String toString() {
    return String.valueOf(value);
  }

  @Override
  protected String toStringNeg() {
    return toString();
  }

  @Override
  public boolean hasZero() {
    return value == 0;
  }

  @Override
  public int noNegConstants() {

    return Math.abs(value);
  }

}

class Neg extends Expr {

  private final Expr operand;

  public Neg(Expr operand) {
    this.operand = operand;
  }

  @Override
  public int eval() {
    return -operand.eval();
  }

  @Override
  public String toString() {
    return "- " + operand.toStringNeg();
  }

  @Override
  public boolean hasZero() {
    return operand.hasZero();
  }

  @Override
  public int noNegConstants() {
    return Math.abs(operand.eval());
  }

}

class Add extends Expr {

  private final Expr operand1;
  private final Expr operand2;

  public Add(Expr operand1, Expr operand2) {
    this.operand1 = operand1;
    this.operand2 = operand2;
  }

  @Override
  public int eval() {
    return operand1.eval() + operand2.eval();
  }

  @Override
  public String toString() {
    final String plus = operand2 instanceof Neg ? " " : " + ";
    return operand1.toString() + plus + operand2.toString();
  }

  @Override
  public boolean hasZero() {
    return operand1.hasZero() || operand2.hasZero();
  }

  @Override
  public int noNegConstants() {
    return Math.abs(operand1.eval()+operand2.eval());
  }

  class Mult extends Expr{
    
    private final Expr operand1;
    private final Expr operand2;

    public Mult(Expr operand1, Expr operand2){
      this.operand1=operand1;
      this.operand2=operand2;
    }

    @Override
    public int eval() {
      return operand1.eval()*operand2.eval();
    }

    @Override
    public String toString() {
      return operand1.eval()*operand2.eval() + " : "+operand1.eval()+"*"+operand2.eval();
    }

    @Override
    public boolean hasZero() {
      return operand1.hasZero() || operand2.hasZero();
    }

    @Override
    public int noNegConstants() {
      return Math.abs(operand1.eval()*operand2.eval());
    }



  }

}

File: Haskell/ue_7/ue_7.hs
{- Codevorlage zum 7. Foliensatz: Arithmetische Ausdrücke
   Alternative Programmierparadigmen 2025
   Prof. Baltasar Trancón Widemann -}

data Expr = Constant Int
          | Neg Expr
          | Add Expr Expr
          | Mult Expr Expr
  deriving (Eq, Show)


eval :: Expr -> Int
eval e = case e of
  Constant c -> c
  Neg a      -> -(eval a)
  Add a b    -> eval a + eval b
  Mult a b   -> eval a * eval b
  

toString :: Expr -> String
toString e = case e of
  Constant c -> show c
  Neg  a     -> "- " ++ toStringNeg a
  Add  a b   -> toString a ++ plus ++ toString b
                  where plus = case b of
                          Neg _ -> " "
                          _     -> " + "
  
  Mult a b -> toString a ++ plus ++ toString b
                  where plus = case b of
                          Neg _ -> " "
                          _     -> " * "
  where
    toStringNeg e = case e of
      Constant _ -> toString e
      _   -> "(" ++ toString e ++ ")"

hasZero :: Expr -> Bool
hasZero e = case e of
  Constant c -> c == 0
  Neg a      -> hasZero a
  Add a b    -> hasZero a || hasZero b
  Mult a b -> hasZero a || hasZero b

noNegConstants :: Expr -> Expr
noNegConstants e = case e of
  Constant c -> if c < 0 then Neg(Constant (abs c))
                else Constant c  
                
  Neg c -> Neg (noNegConstants c)

  Add a b -> Add (noNegConstants a) (noNegConstants b) 
  Mult a b -> Mult (noNegConstants a) (noNegConstants b)



  --habs im Endeffekt nur für Hs gemacht. Dort war die neue Funktion aufwändiger als die Datenvariante. Ich kann mir vorstellen dass es bei ähnlich ist. jedoch bei hs einfacher